# Wanderbricks — Test the Serving Endpoint

Call the deployed `wanderbricks-weather-serve` endpoint with a realistic payload built from the actual feature table. Measures round-trip latency (includes network — the endpoint UI's Metrics tab shows the server-side p50/p95/p99 latency).

**Prereqs:** the DLT pipeline has run, a model version is registered, and the serving endpoint is deployed (rename the `.example` resource and `databricks bundle deploy`).

**Dev-mode name:** resources are prefixed `dev_<user>_` in the dev
target — the endpoint is `dev_iraonfridays_wanderbricks-weather-serve`
here; in prod it is `wanderbricks-weather-serve`.

In [1]:
# --- Bootstrap: workspace OR local (Databricks Connect + SDK) ---
try:
    spark
except NameError:
    from databricks.connect import DatabricksSession
    spark = DatabricksSession.builder.profile("irajput").serverless().getOrCreate()

try:
    dbutils
except NameError:
    from databricks.sdk import WorkspaceClient
    _w = WorkspaceClient(profile="irajput")
    dbutils = _w.dbutils

# display() exists natively in the workspace; fall back to show() locally
try:
    display
except NameError:
    def display(df, *a, **k):
        df.show(20)

# Widgets: workspace-native, SDK-emulated, or absent -> always safe
def _get_widget(name, default=""):
    try:
        dbutils.widgets.text(name, default)
        return dbutils.widgets.get(name).strip()
    except Exception:
        return default

CATALOG = _get_widget("catalog", "workspace")
SCHEMA = _get_widget("schema", "iraonfridays")
print(f"Connected to Databricks! Target: {CATALOG}.{SCHEMA}")

ENDPOINT = _get_widget("endpoint", "dev_iraonfridays_wanderbricks-weather-serve")

Box(children=(Label(value='catalog'), Text(value='workspace')))

Box(children=(Label(value='schema'), Text(value='iraonfridays')))

Connected to Databricks! Target: workspace.iraonfridays


Box(children=(Label(value='endpoint'), Text(value='dev_iraonfridays_wanderbricks-weather-serve')))

## Build a realistic request payload

In [2]:
import pyspark.sql.functions as F

# One real feature row per station (drop non-model columns)
feats = (
    spark.table(f"{CATALOG}.{SCHEMA}.weather_features")
    .drop("station", "date", "temp_c")
)
sample = feats.dropna().limit(1).toPandas().to_dict(orient="records")
print(sample)
assert sample, "weather_features is empty - run the DLT pipeline first"

[{'day_of_week': 2, 'day_of_year': 34, 'month': 2, 'is_weekend': 0, 'temp_lag_1': 5.0, 'temp_lag_7': 3.67, 'temp_lag_14': 7.67, 'temp_lag_28': -2.0, 'temp_rollmean_7': 4.6899999999999995, 'temp_rollmean_14': 4.225714285714285}]


## Call the endpoint and time it

In [3]:
import requests, time, json

# Workspace: notebook context. Locally: SDK host + CLI token cache.
try:
    ctx = dbutils.notebook.entry_point.getDbutils().notebook().getContext()
    host = ctx.apiUrl().get()
    token = ctx.apiToken().get()
except Exception:
    from databricks.sdk import WorkspaceClient
    import subprocess
    host = WorkspaceClient(profile="irajput").config.host
    token = json.loads(subprocess.check_output(["databricks", "auth", "token", "irajput"]))["access_token"]

payload = {"dataframe_records": sample}
t0 = time.perf_counter()
resp = requests.post(
    f"{host}/serving-endpoints/{ENDPOINT}/invocations",
    headers={"Authorization": f"Bearer {token}"},
    json=payload,
    timeout=60,
)
elapsed = time.perf_counter() - t0
print(f"HTTP {resp.status_code} · round-trip {elapsed:.4f}s")
if resp.status_code == 200:
    print(json.dumps(resp.json(), indent=2))
else:
    print(resp.text)

I0830 18:01:44.580919 6040833 ev_poll_posix.cc:593] FD from fork parent still in poll list: fd(95, generation: 1)
I0830 18:01:46.094261 6040910 ev_poll_posix.cc:593] FD from fork parent still in poll list: fd(95, generation: 1)


HTTP 200 · round-trip 34.7290s
{
  "predictions": [
    4.318821907043457
  ]
}


In [6]:
from databricks.sdk import WorkspaceClient

w = WorkspaceClient(profile="irajput")

# Sample input payload matching your feature schema
payload = {
    "dataframe_records": [
        {
            "day_of_week": 2,
            "day_of_year": 243,
            "month": 8,
            "is_weekend": 0,
            "temp_lag_1": 21.4,
            "temp_lag_7": 20.8,
            "temp_lag_14": 19.5,
            "temp_lag_28": 18.9,
            "temp_rollmean_7": 20.9,
            "temp_rollmean_14": 20.2,
        }
    ]
}

# Query using the exact endpoint name
response = w.serving_endpoints.query(
    name="dev_iraonfridays_wanderbricks-weather-serve",
    dataframe_records=payload["dataframe_records"],
)

print(f"Predicted Temperature (°C): {response.predictions}")


Predicted Temperature (°C): [22.605262756347656]


In [7]:
spark.sql(f"""
    SELECT run_date, rows, inference_seconds 
    FROM {CATALOG}.{SCHEMA}.scoring_metrics 
    ORDER BY run_date DESC
""").show()


+----------+-----+-----------------+
|  run_date| rows|inference_seconds|
+----------+-----+-----------------+
|2026-08-30|11329|           0.0491|
+----------+-----+-----------------+



In [8]:
spark.table(f"{CATALOG}.{SCHEMA}.scoring_metrics").printSchema()


root
 |-- run_date: string (nullable = true)
 |-- rows: long (nullable = true)
 |-- inference_seconds: double (nullable = true)



In [10]:
# In your notebook or SQL Editor:
spark.sql(f"""
    SELECT 
        run_date, 
        rows, 
        inference_seconds,
        ROUND(rows / inference_seconds, 1) AS rows_per_second
    FROM {CATALOG}.{SCHEMA}.scoring_metrics 
    ORDER BY run_date DESC
""").show()


+----------+-----+-----------------+---------------+
|  run_date| rows|inference_seconds|rows_per_second|
+----------+-----+-----------------+---------------+
|2026-08-30|11329|           0.0491|       230733.2|
+----------+-----+-----------------+---------------+



This result is your **Batch Inference Performance benchmark** from running [ml/06_score.py](file:///Users/irajput/databricks-forecasting/ml/06_score.py). 

Here is what each number means:

---

### Column Breakdown:

| Column | Value | Meaning |
| :--- | :--- | :--- |
| **`run_date`** | **`2026-08-30`** | The date the automated batch scoring job executed. |
| **`rows`** | **`11,329`** | The total number of station-day records that were scored. |
| **`inference_seconds`** | **`0.0491 s`** | The total model calculation time (**~49 milliseconds**). |
| **`rows_per_second`** | **`230,733.2`** | The scoring throughput (**over 230,000 forecasts per second**). |

---

### What this tells you about your pipeline:

1. **Blazing Fast Batch Scoring:**
   - Your trained XGBoost model scored the **entire daily dataset (11,329 forecasts) in less than 50 milliseconds**.
   - This is why gradient-boosted trees (XGBoost) are an industry favorite for tabular forecasting: they deliver high accuracy with near-instantaneous batch throughput.

2. **Resource Efficiency & Low Cloud Cost:**
   - Because scoring takes under a tenth of a second, your Databricks ephemeral job cluster spends virtually zero compute time on inference, keeping your cloud bill minimal.

3. **Performance Drift Monitoring Baseline:**
   - This row serves as your **performance baseline**. As your historical data grows over weeks and months (e.g., from 11k rows $\to$ 1 million rows), you can monitor `inference_seconds` and `rows_per_second` to ensure your batch pipeline continues to scale linearly without memory bottlenecks or latency regressions.

## Reading the results

- **Round-trip** here includes network + endpoint dispatch; the server-side number is in the endpoint's **Metrics tab** (p50/p95/p99 latency, requests/sec).
- **First call after idle** may be slower — `scale_to_zero_enabled: true` means cold starts; keep it on for cost, off for guaranteed latency.
- The response should match the model signature (a `predictions` array of temperatures in °C).

Compare with the batch path: `SELECT * FROM <schema>.scoring_metrics ORDER BY run_date` shows `inference_seconds` from `06_score.py`.